File: android/app/src/main/java/app/releaf/mobile/data/sync/CanonicalJson.kt
/*
 * CanonicalJson.kt
 *
 * Canonical JSON serializer for Drive checksumming.
 *
 * Per `docs/DRIVE_SCHEMA.md` §"Canonical JSON for checksumming":
 *   1. UTF-8 encoding.
 *   2. Keys sorted lexicographically within every object.
 *   3. No insignificant whitespace (no newlines, no trailing spaces).
 *   4. Numbers serialize as JSON integers when integer-valued, otherwise
 *      shortest round-tripping decimal.
 *   5. Strings use JSON default escaping.
 *
 * iOS and Android must produce byte-identical output on identical input.
 * A cross-platform fixture lives at `design-system/fixtures/canonical-
 * json-fixture.json`; both platforms' test suites canonicalize it and
 * assert the same SHA-256 bytes.
 *
 * Callers pass in a kotlinx `JsonElement` — the caller is responsible
 * for constructing the element using kotlinx.serialization (which emits
 * `Long` as integer and `Double` as shortest decimal). The canonicalizer
 * never invents numeric forms; it sorts keys, strips whitespace, and
 * re-escapes strings.
 */

package app.releaf.mobile.data.sync

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

object CanonicalJson {

    /** Encode an element to canonical UTF-8 bytes. */
    fun encodeToBytes(element: JsonElement): ByteArray =
        encodeToString(element).toByteArray(Charsets.UTF_8)

    /** Encode an element to a canonical JSON string. */
    fun encodeToString(element: JsonElement): String =
        buildString { writeElement(this, element) }

    private fun writeElement(sb: StringBuilder, e: JsonElement) {
        when (e) {
            is JsonNull      -> sb.append("null")
            is JsonPrimitive -> writePrimitive(sb, e)
            is JsonArray     -> {
                sb.append('[')
                var first = true
                for (child in e) {
                    if (!first) sb.append(',')
                    writeElement(sb, child)
                    first = false
                }
                sb.append(']')
            }
            is JsonObject    -> {
                sb.append('{')
                var first = true
                for (key in e.keys.sorted()) {
                    if (!first) sb.append(',')
                    writeString(sb, key)
                    sb.append(':')
                    writeElement(sb, e.getValue(key))
                    first = false
                }
                sb.append('}')
            }
        }
    }

    private fun writePrimitive(sb: StringBuilder, p: JsonPrimitive) {
        if (p.isString) {
            writeString(sb, p.content)
        } else {
            // Numeric / boolean / null-as-primitive — emit kotlinx's
            // stringified form verbatim. Callers must pass numbers via
            // the typed `JsonPrimitive(Long)` / `JsonPrimitive(Double)`
            // constructors so the text matches what a number literal
            // would produce on the other platform.
            sb.append(p.content)
        }
    }

    /**
     * Default JSON string escaping. Matches the set RFC 8259 requires
     * and what Apple's JSONSerialization emits under .sortedKeys, so
     * Android and iOS agree byte-for-byte on escapes.
     */
    private fun writeString(sb: StringBuilder, s: String) {
        sb.append('"')
        for (c in s) {
            when (c) {
                '"'      -> sb.append("\\\"")
                '\\'     -> sb.append("\\\\")
                '\b'     -> sb.append("\\b")
                '\u000c' -> sb.append("\\f")
                '\n'     -> sb.append("\\n")
                '\r'     -> sb.append("\\r")
                '\t'     -> sb.append("\\t")
                else     -> if (c.code < 0x20) {
                    sb.append("\\u")
                    sb.append("%04x".format(c.code))
                } else {
                    sb.append(c)
                }
            }
        }
        sb.append('"')
    }
}

/**
 * Hex-encoded SHA-256 of the canonical-JSON bytes for [element].
 * Convenience — every caller that checksums a payload wants the
 * same pipeline (canonicalize → SHA-256 → lowercase hex).
 */
fun sha256Hex(bytes: ByteArray): String {
    val digest = java.security.MessageDigest.getInstance("SHA-256").digest(bytes)
    return buildString(digest.size * 2) {
        for (b in digest) {
            val v = b.toInt() and 0xff
            append(HEX[v ushr 4])
            append(HEX[v and 0x0f])
        }
    }
}

private const val HEX = "0123456789abcdef"

File: shared/ios/ReleafCore/Sources/ReleafCoreSync/CanonicalJson.swift
/*
 * CanonicalJson.swift
 *
 * Canonical JSON serializer for Drive checksumming.
 *
 * Per `docs/DRIVE_SCHEMA.md` §"Canonical JSON for checksumming":
 *   1. UTF-8 encoding.
 *   2. Keys sorted lexicographically within every object.
 *   3. No insignificant whitespace.
 *   4. Numbers emit integers when integer-valued, otherwise shortest
 *      round-trip decimal.
 *   5. Default JSON string escaping.
 *
 * iOS and Android must produce byte-identical output on identical
 * input. A cross-platform fixture at
 * `design-system/fixtures/canonical-json-fixture.json` feeds both
 * platforms' test suites; they assert the same SHA-256 bytes.
 *
 * Strategy: take a Foundation JSON tree (NSDictionary / NSArray /
 * NSNumber / NSString / NSNull) and walk it manually. We don't rely on
 * JSONSerialization's .sortedKeys because it doesn't give bit-for-bit
 * control over number formatting or escape selection.
 */

import Foundation
import CommonCrypto

public enum CanonicalJson {

    /// Encode a Foundation JSON tree to canonical UTF-8 bytes.
    public static func encodeToData(_ value: Any) -> Data {
        var out = String()
        write(value, into: &out)
        return Data(out.utf8)
    }

    /// Encode a Foundation JSON tree to a canonical JSON string.
    public static func encodeToString(_ value: Any) -> String {
        var out = String()
        write(value, into: &out)
        return out
    }

    /// Encode an `Encodable` via JSONEncoder, then walk the resulting
    /// tree through the canonicalizer. Convenience for payload types
    /// defined with `Codable`.
    public static func encodeToData<T: Encodable>(encodable: T) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = []
        let data = try encoder.encode(encodable)
        let tree = try JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
        return encodeToData(tree)
    }

    // MARK: - Implementation

    private static func write(_ value: Any, into sb: inout String) {
        if value is NSNull { sb.append("null"); return }

        switch value {
        case let dict as [String: Any]:
            writeObject(dict, into: &sb)

        case let nsdict as NSDictionary:
            var map: [String: Any] = [:]
            for case let (k as String, v) in nsdict { map[k] = v }
            writeObject(map, into: &sb)

        case let array as [Any]:
            writeArray(array, into: &sb)

        case let nsarr as NSArray:
            writeArray(Array(nsarr), into: &sb)

        case let s as String:
            writeString(s, into: &sb)

        case let n as NSNumber:
            writeNumber(n, into: &sb)

        case let b as Bool:
            sb.append(b ? "true" : "false")

        case let i as Int:
            sb.append(String(i))

        case let i as Int64:
            sb.append(String(i))

        case let d as Double:
            sb.append(shortestDecimal(d))

        default:
            sb.append("null")
        }
    }

    private static func writeObject(_ dict: [String: Any], into sb: inout String) {
        sb.append("{")
        var first = true
        for key in dict.keys.sorted() {
            if !first { sb.append(",") }
            writeString(key, into: &sb)
            sb.append(":")
            write(dict[key] as Any, into: &sb)
            first = false
        }
        sb.append("}")
    }

    private static func writeArray(_ arr: [Any], into sb: inout String) {
        sb.append("[")
        var first = true
        for v in arr {
            if !first { sb.append(",") }
            write(v, into: &sb)
            first = false
        }
        sb.append("]")
    }

    /// `NSNumber` double-duties as bool / int / double out of Foundation
    /// JSON decode. Disambiguate via the Objective-C type encoding —
    /// booleans come back as `@"c"` (signed char) on iOS.
    private static func writeNumber(_ n: NSNumber, into sb: inout String) {
        let type = String(cString: n.objCType)
        if type == "c" || type == "B" {
            sb.append(n.boolValue ? "true" : "false")
            return
        }
        // Integer-valued doubles emit as integers per spec §4.
        // Use Int64(exactly:) — the obvious `d <= Double(Int64.max)`
        // bound check is a trap: Double(Int64.max) rounds UP to
        // 9223372036854775808.0, so values right at that boundary
        // pass the bound then crash on Int64(d).
        let d = n.doubleValue
        if let i = Int64(exactly: d) {
            sb.append(String(i))
        } else {
            sb.append(shortestDecimal(d))
        }
    }

    /// Swift's `String(d)` is round-trip-safe for finite doubles and
    /// matches Kotlin's `Double.toString()` on finite values.
    private static func shortestDecimal(_ d: Double) -> String {
        if d.isNaN || d.isInfinite { return "null" }
        return String(d)
    }

    private static func writeString(_ s: String, into sb: inout String) {
        sb.append("\"")
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"":     sb.append("\\\"")
            case "\\":     sb.append("\\\\")
            case "\u{08}": sb.append("\\b")
            case "\u{0C}": sb.append("\\f")
            case "\n":     sb.append("\\n")
            case "\r":     sb.append("\\r")
            case "\t":     sb.append("\\t")
            default:
                if scalar.value < 0x20 {
                    sb.append(String(format: "\\u%04x", scalar.value))
                } else {
                    sb.append(Character(scalar))
                }
            }
        }
        sb.append("\"")
    }
}

/// Hex-encoded SHA-256 of the given bytes. Every caller that hashes a
/// payload wants the same pipeline (canonicalize → hash → lowercase hex).
public func sha256Hex(_ data: Data) -> String {
    var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    data.withUnsafeBytes { bytes in
        _ = CC_SHA256(bytes.baseAddress, CC_LONG(data.count), &digest)
    }
    let hex = "0123456789abcdef"
    var out = String()
    out.reserveCapacity(digest.count * 2)
    for b in digest {
        let hi = Int(b >> 4)
        let lo = Int(b & 0x0f)
        out.append(hex[hex.index(hex.startIndex, offsetBy: hi)])
        out.append(hex[hex.index(hex.startIndex, offsetBy: lo)])
    }
    return out
}
